File: Java/ElGamal.java
import java.io.*;
import java.math.BigInteger;

public class ElGamal extends PublicKeyCryptosystem {
    protected BigInteger primElement;

    public ElGamal() {
        super();
        setModulus();
        setPrimElement();
        setPrivateKey();
        setPublicKey();
    }
    public ElGamal(BigInteger privateKey) {
        super(privateKey);
        setModulus();
        setPrimElement();
        setPublicKey();
    }
    public ElGamal(int privateKey) {
        super(privateKey);
        setModulus();
        setPrimElement();
        setPublicKey();
    }
    public ElGamal(BigInteger modulus, BigInteger privateKey, BigInteger primElement) {
        super(modulus, privateKey);
        setPrimElement(primElement);
        setPublicKey();
    }
    public ElGamal(int modulus, int privateKey, int primElement) {
        super(modulus, privateKey);
        setPrimElement(primElement);
        setPublicKey();
    }

    protected void setModulus() {
        try {
            modulus = BigInteger.valueOf(super.genRandPrime().getFirst());
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
    protected void setPrivateKey() {
        BigInteger randPrivateKey = new BigInteger(super.genRandValue(2, modulus.longValue()).toString());
        this.privateKey = new BigInteger(randPrivateKey.toString());
    }
    protected void setPrivateKey(BigInteger privateKey) {
        this.privateKey = new BigInteger(privateKey.toString());
    }
    protected void setPrivateKey(int privateKey) {
        this.privateKey = new BigInteger(String.valueOf(privateKey));
    }
    protected void setPublicKey() {
        if (primElement != null && modulus != null && privateKey != null) {
            this.publicKey = new BigInteger(primElement.modPow(privateKey, modulus).toString());
        }
    }

    protected void setPrimElement() {
        primElement = new BigInteger("2");
    }
    protected void setPrimElement(BigInteger primElement) {
        this.primElement = new BigInteger(primElement.toString());
    }
    protected void setPrimElement(int primElement) {
        this.primElement = new BigInteger(String.valueOf(primElement));
    }

    public BigInteger getPrimElement() {
        return primElement;
    }

    public Pair<BigInteger> encrypt(BigInteger plainText, BigInteger randomKey) {
        BigInteger cipherText1 = primElement.modPow(randomKey, modulus);
        BigInteger cipherText2 = publicKey.modPow(randomKey, modulus);
        cipherText2 = cipherText2.multiply(plainText);
        cipherText2 = cipherText2.mod(modulus);

        Pair<BigInteger> encryptedText = new Pair<BigInteger>(cipherText1, cipherText2);

        return encryptedText;
    }
    public Pair<BigInteger> timedEncrypt(BigInteger plainText, BigInteger randomKey, File output, Character delim) throws IOException {
        int bitLength = privateKey.bitLength();

        long startTime = System.nanoTime();
        Pair<BigInteger> encryptedText = encrypt(plainText, randomKey);
        long endTime = System.nanoTime();

        output.createNewFile();
        FileWriter writer = new FileWriter(output.getName(), true);
        PrintWriter printer = new PrintWriter(writer);
        printer.println(bitLength + delim.toString() + (endTime - startTime));
        writer.close();
        printer.close();

        return encryptedText;
    }
    public BigInteger decrypt(Pair<BigInteger> cipherText) {
        BigInteger x = cipherText.getFirst();
        x = x.modPow(privateKey, modulus);
        x = x.modInverse(modulus);
        x = x.multiply(cipherText.getSecond());
        return x.mod(modulus);
    }
    public BigInteger timedDecrypt(Pair<BigInteger> cipherText, File output, Character delim) throws IOException {
        int bitLength = privateKey.bitLength();

        long startTime = System.nanoTime();
        BigInteger decryptedText = decrypt(cipherText);
        long endTime = System.nanoTime();

        output.createNewFile();
        FileWriter writer = new FileWriter(output.getName(), true);
        PrintWriter printer = new PrintWriter(writer);
        printer.println(bitLength + delim.toString() + (endTime - startTime));
        writer.close();
        printer.close();

        return decryptedText;
    }
}

File: C++/Sources/ElGamal.cpp
#include "ElGamal.h"

#include "Pair.h"

#include <cmath>
#include <chrono>
#include <ostream>

void ElGamal::setModulus() {
    modulus = genRandPrime().getFirst();
}
void ElGamal::setPrivateKey() {
    privateKey = genRandValue(2, modulus);
}
void ElGamal::setPublicKey() {
    if (primElement != 0 && modulus != 0 && privateKey != 0) {
        this->publicKey = modPow(primElement, privateKey, modulus);
    }
}

void ElGamal::setPrimElement() {
    primElement = 2;
}
void ElGamal::setPrimElement(long long primElement) {
    this->primElement = primElement;
}
void ElGamal::setPrimElement(int primElement) {
    this->primElement = primElement;
}

ElGamal::ElGamal() : PublicKeyCryptosystem() {
    setModulus();
    setPrimElement();
    setPrivateKey();
    setPublicKey();
}
ElGamal::ElGamal(long long privateKey) : PublicKeyCryptosystem(privateKey) {
    setModulus();
    setPrimElement();
    setPublicKey();
}
ElGamal::ElGamal(int privateKey) : PublicKeyCryptosystem(privateKey) {
    setModulus();
    setPrimElement();
    setPublicKey();
}
ElGamal::ElGamal(long long modulus, long long privateKey, long long primElement) : PublicKeyCryptosystem(modulus, privateKey) {
    setPrimElement(primElement);
    setPublicKey();
}
ElGamal::ElGamal(int modulus, int privateKey, int primElement) : PublicKeyCryptosystem(modulus, privateKey) {
    setPrimElement(primElement);
    setPublicKey();
}
ElGamal::~ElGamal() {}

long long ElGamal::getPrimElement() {
    return primElement;
}

Pair<long long> ElGamal::encrypt(long long plainText, long long randomKey) {
    long long cipherText1 = modPow(primElement, randomKey, modulus);
    long long cipherText2 = modPow(publicKey, randomKey, modulus);
    cipherText2 *= plainText;
    cipherText2 = cipherText2 % modulus;
    
    Pair<long long> encryptedText(cipherText1, cipherText2);
    return encryptedText;
}
Pair<long long> ElGamal::timedEncrypt(long long plainText, long long randomKey, std::ostream& os, char delim) {
    int numBits = floor(log2(privateKey)) + 1;
    
    auto startTime = std::chrono::steady_clock::now();
    Pair<long long> encryptedText = encrypt(plainText, randomKey);
    auto endTime = std::chrono::steady_clock::now();

    os << numBits << delim << std::chrono::duration_cast<std::chrono::nanoseconds>(endTime - startTime).count() << std::endl;

    return encryptedText;
}
long long ElGamal::decrypt(Pair<long long> cipherText) {
    long long x = cipherText.getFirst();
    x = modPow(x, privateKey, modulus);
    x = modInverse(x, modulus);
    x = (x * cipherText.getSecond()) % modulus;
    return x;
}
long long ElGamal::timedDecrypt(Pair<long long> cipherText, std::ostream& os, char delim) {
    int numBits = floor(log2(privateKey)) + 1;
    
    auto startTime = std::chrono::high_resolution_clock::now();
    long long decryptedText = decrypt(cipherText);
    auto endTime = std::chrono::high_resolution_clock::now();

    os << numBits << delim << std::chrono::duration_cast<std::chrono::nanoseconds>(endTime - startTime).count() << std::endl;
    
    return decryptedText;
}
